src/gdb_packet.c

```diff
@@ -31,119 +31,220 @@
 
 #include <stdarg.h>
 
+typedef enum packet_state {
+	PACKET_IDLE,
+	PACKET_GDB_CAPTURE,
+	PACKET_GDB_ESCAPE,
+	PACKET_GDB_CHECKSUM_UPPER,
+	PACKET_GDB_CHECKSUM_LOWER,
+	PACKET_GDB_COMPLETE,
+} packet_state_e;
+
+static bool noackmode = false;
+
+void gdb_set_noackmode(bool enable)
+{
+	/* 
+	 * If we were asked to disable no ack mode, and it was previously enabled,
+	 * it might mean we got a paket we determined to be the first of a new
+	 * GDB session, and as such it was not acknoledged (before GDB enabled no ack mode),
+	 * better late than never.
+	 * 
+	 * If we were asked after the connection was terminated, sending the ack have no effect.
+	 */
+	if (enable == false && noackmode == true) {
+		gdb_if_putchar(GDB_PACKET_ACK, 1U);
+	}
+
+	DEBUG_GDB("%s no ack mode\n", enable ? "Enable" : "Disable");
+
+	noackmode = enable;
+}
+
+packet_state_e consume_remote_packet(char *const packet, const size_t size)
+{
+	/* We got what looks like probably a remote control packet */
+
+#if PC_HOSTED == 0
+	size_t offset = 0;
+
+	while (true) {
+		/* Consume bytes until we either have a complete remote control packet or have to leave this mode */
+		const char rx_char = gdb_if_getchar();
+
+		switch (rx_char) {
+		case '\x04':
+			/* EOT (end of transmission) - connection was closed */
+			return PACKET_IDLE;
+
+		case REMOTE_SOM:
+			/* Oh dear, restart remote packet capture */
+			offset = 0;
+			break;
+
+		case REMOTE_EOM:
+			/* Complete packet for processing */
+
+			/* Null terminate packet */
+			packet[offset] = '\0';
+
+			/* Handle packet */
+			remote_packet_process(offset, packet);
+
+			/* Restart packet capture */
+			return PACKET_IDLE;
+
+		case GDB_PACKET_START:
+			/* A 'real' gdb packet, best stop squatting now */
+			return PACKET_GDB_CAPTURE;
+
+		default:
+			if (offset < size)
+				packet[offset++] = rx_char;
+			else
+				/* Buffer overflow, restart packet capture */
+				return PACKET_IDLE;
+		}
+	}
+#else
+	(void)packet;
+	(void)size;
+
+	/* Hosted builds ignore remote control packets */
+	return PACKET_IDLE;
+#endif
+}
+
 size_t gdb_getpacket(char *const packet, const size_t size)
 {
-	unsigned char csum;
-	char recv_csum[3];
+	packet_state_e state = PACKET_IDLE; /* State of the packet capture */
+
 	size_t offset = 0;
+	uint8_t rx_checksum = 0, checksum = 0;
 
 	while (true) {
-		/* Wait for packet start */
-		do {
-			/*
-			 * Spin waiting for a start of packet character - either a gdb
-             * start ('$') or a BMP remote packet start ('!').
-			 */
-			do {
-				/* Smells like bad code */
-				packet[0] = gdb_if_getchar();
-				if (packet[0] == '\x04')
-					return 1;
-			} while (packet[0] != '$' && packet[0] != REMOTE_SOM);
-#if PC_HOSTED == 0
-			if (packet[0] == REMOTE_SOM) {
-				/* This is probably a remote control packet - get and handle it */
+		// const char rx_char = state != PACKET_GDB_COMPLETE ? gdb_if_getchar() : '\0';
```

Comment:
Please drop the commented out line of code

Reply:
This is part of a temporary commit to get things works, this is why I noted "theres still an issue" this wont be part of the final submission